pkg/providers/providers.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2024
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package providers
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	hmc "github.com/Mirantis/hmc/api/v1alpha1"
+	"github.com/Mirantis/hmc/pkg/credspropagation"
+	"github.com/Mirantis/hmc/pkg/providers/aws"
+)
+
+const (
+	InfraPrefix    = "infrastructure-"
+	ProviderPrefix = "cluster-api-provider-"
+)
+
+var (
+	mu sync.RWMutex
+
+	providers = []hmc.Provider{
+		{
+			Name: hmc.ProviderK0smotronName,
+		},
+		{
+			Name: hmc.ProviderSveltosName,
+		},
+	}
+
+	registry map[string]ProviderModule
+)
+
+type ProviderModule interface {
+	GetName() string
+	GetTitleName() string
+	GetClusterGVK() schema.GroupVersionKind
+	GetClusterIdentityKind() []string
+	CredentialPropagationFunc() func(
+		ctx context.Context,
+		propnCfg *credspropagation.PropagationCfg,
+		l logr.Logger,
```

Comment:
Not sure what you mean about context. Function itself is called here https://github.com/s3rj1k/hmc/blob/provider-modules/internal/controller/managedcluster_controller.go#L722

Removing logger from signature would force fetching logger as many time as there are providers.

Technically possible to remove.

Reply:
Yep, I made a mistake. The instrumentation of the logger's key is indeed in the [logr](https://github.com/go-logr/logr/blob/master/context.go#L22) package itself, so it'd be possible. I think having hundreds of providers won't affect efficiency too much, and we only expect a handful of providers.

Reply:
```
func (p *Provider) CredentialPropagationFunc() func(
	ctx context.Context,
	propnCfg *credspropagation.PropagationCfg,
	l logr.Logger,
) (enabled bool, err error) {
	return func(
		ctx context.Context,
		propnCfg *credspropagation.PropagationCfg,
	) (enabled bool, err error) {
		l := ctrl.LoggerFrom(ctx) // ctrl -> "sigs.k8s.io/controller-runtime"
		l.Info(p.GetTitleName() + " creds propagation start")
		enabled, err = true, PropagateSecrets(ctx, propnCfg)
		return enabled, err
	}
}
```

Adds `"sigs.k8s.io/controller-runtime"` dependency for plugin, not sure that this is a good idea.

To avoid that we would need to copy this function out of controller-runtime
```
// FromContext returns a logger with predefined values from a context.Context.
func FromContext(ctx context.Context, keysAndValues ...interface{}) logr.Logger {
	log := Log
	if ctx != nil {
		if logger, err := logr.FromContext(ctx); err == nil {
			log = logger
		}
	}
	return log.WithValues(keysAndValues...)
}
```